cpp/src/parquet/encoding-test.cc

```diff
@@ -314,6 +325,258 @@ TEST(TestDictionaryEncoding, CannotDictDecodeBoolean) {
   ASSERT_THROW(MakeDictDecoder<BooleanType>(nullptr), ParquetException);
 }
 
+// ----------------------------------------------------------------------
+// Shared arrow builder decode tests
+class TestDecodeArrow : public ::testing::TestWithParam<const char*> {
+ public:
+  void SetUp() override {
+    InitFromJSON(GetParam());
+    SetupEncoderDecoder();
+  }
+
+  void InitFromJSON(const char* json) {
+    // Use input JSON to initialize the dense array
+    expected_dense_ = ::arrow::ArrayFromJSON(::arrow::binary(), json);
+    num_values_ = static_cast<int>(expected_dense_->length());
+    null_count_ = static_cast<int>(expected_dense_->null_count());
+    valid_bits_ = expected_dense_->null_bitmap()->data();
+
+    // Build both binary and string dictionary arrays from the dense array.
+    BuildDict<::arrow::BinaryDictionaryBuilder>(&expected_bin_dict_);
+    BuildDict<::arrow::StringDictionaryBuilder>(&expected_str_dict_);
+
+    // Initialize input_data_ for the encoder from the expected_array_ values
+    const auto& binary_array = static_cast<const ::arrow::BinaryArray&>(*expected_dense_);
+    input_data_.reserve(binary_array.length());
+
+    for (int64_t i = 0; i < binary_array.length(); ++i) {
+      auto view = binary_array.GetView(i);
+      input_data_.emplace_back(static_cast<uint32_t>(view.length()),
+                               reinterpret_cast<const uint8_t*>(view.data()));
+    }
+  }
+
+  // Builds a dictionary encoded array from the dense array using BuilderType
+  template <typename BuilderType>
+  void BuildDict(std::shared_ptr<::arrow::Array>* result) {
+    const auto& binary_array = static_cast<const ::arrow::BinaryArray&>(*expected_dense_);
+    BuilderType builder(default_memory_pool());
+
+    for (int64_t i = 0; i < binary_array.length(); ++i) {
+      if (binary_array.IsNull(i)) {
+        ASSERT_OK(builder.AppendNull());
+      } else {
+        ASSERT_OK(builder.Append(binary_array.GetView(i)));
+      }
+    }
```

Comment:
Can you use `AppendArray` instead? https://github.com/apache/arrow/blob/master/cpp/src/arrow/array/builder_dict.cc#L243

Reply:
Yes `AppendArray` is neater.